cf-custom-resources/lib/alb-rule-priority-generator.js

```diff
@@ -0,0 +1,141 @@
+ * @param {string} physicalResourceId CloudFormation physical resource ID
+ * @param {object} [responseData] arbitrary response data object
+ * @param {string} [reason] reason for failure, if any, to convey to the user
+ * @returns {Promise} Promise that is resolved on success, or rejected on connection error or HTTP error response
+ */
+let report = function (event, context, responseStatus, physicalResourceId, responseData, reason) {
+    return new Promise((resolve, reject) => {
+        const https = require('https');
+        const {
+            URL
+        } = require('url');
+
+        var responseBody = JSON.stringify({
+            Status: responseStatus,
+            Reason: reason,
+            PhysicalResourceId: physicalResourceId || context.logStreamName,
+            StackId: event.StackId,
+            RequestId: event.RequestId,
+            LogicalResourceId: event.LogicalResourceId,
+            Data: responseData
+        });
+
+        const parsedUrl = new URL(event.ResponseURL || defaultResponseURL);
+        const options = {
+            hostname: parsedUrl.hostname,
+            port: 443,
+            path: parsedUrl.pathname + parsedUrl.search,
+            method: 'PUT',
+            headers: {
+                'Content-Type': '',
+                'Content-Length': responseBody.length
+            }
+        };
+
+        https.request(options)
+            .on('error', reject)
+            .on('response', res => {
+                res.resume();
+                if (res.statusCode >= 400) {
+                    reject(new Error(`Error ${res.statusCode}: ${res.statusMessage}`));
+                } else {
+                    resolve();
+                }
+            })
+            .end(responseBody, 'utf8');
+    });
+};
+
+/**
+ * Lists all the existing rules for a ALB Listener, finds the max of their
+ * priorities, and then returns max + 1.
+ *
+ * @param {string} listenerArn the ARN of the ALB listener.
+
+ * @returns {number} The next available ALB listener rule priority.
+ */
+const calculateNextRulePriority = async function (listenerArn) {
+    var elb = new aws.ELBv2();
+    // Grab all the rules for this listener
+    var marker;
+    var rules = []
+    do {
+        const rulesResponse = await elb.describeRules({
+            ListenerArn: listenerArn,
+            Marker: marker
+        }).promise();
+
+        rules = rules.concat(rulesResponse.Rules)
+        marker = rulesResponse.NextMarker;
+    } while (marker)
+
+    let nextRulePriority = 1;
+    if (rules.length > 0) {
+        // Take the max rule priority, and add 1 to it.
+        const rulePriorities = rules.map(rule => {
+            if (rule.Priority === "default") {
+                // We treat the default rule as having priority 0
+                return 0
+            }
+            return parseInt(rule.Priority);
+        });
+        const max = Math.max(...rulePriorities);
```

Comment:
@kohidave if a rule is deleted we are not grabbing the "missing" priority number, instead we are always incrementing. I guess that's also fine?